### ISSUE TITLE
Replays of story scenarios do not allow normal replay controls

### ISSUE THREAD

**Issue by Ashmyr** [wesnoth/wesnoth#7122]
### Game and System Information

- **Version: 1.16**
- **Downloaded from: The Website**
- **Build info:** (confused on what this is asking)
- **OS: Windows 11**

Got a new computer. Old Bug Report was dismissed because of the OS and I wasn't playing on a supported version. So here is a new bug report about the same thing, on a new computer, on a supported version, lol.

### Description of the bug

Cutscene replays work identically to gz files. They don't allow you to inspect the map, see units you might have missed, etc. It's a bit irritating. If you delete the theme code and linger_mode, then it works fine, but the theme is pretty cool and is there for a reason, so I assume that's not a solution, lol

### Steps to reproduce the behavior

1. Load cutscene scenario
2. Start next scenario
3. Load cutscene replay

### Expected behavior

The replay should work like a replay. You should be able to pause before the dialogue starts, check out the map and see what units are there and their stats.

**Comment by knyghtmare** [comment]
> check out the map and see what units are there and their stats.

I think you have mixed up a non-combat scenario with a cutscene. Can you give a mainline scenario example of where this is happening and the intended state?

**Comment by Ashmyr** [comment]
I'm playing on 1.16, TRoW. Southbay in Winter and Return to Oldwood both have this bug.

**Comment by Wedge009** [comment]
Both of those are story/non-combat scenarios. I'm fairly sure they are working as intended.

**Comment by Wedge009** [comment]
Closing in favour of #7123.

**Comment by soliton-** [comment]
#7123 is about not saving replays. This issue is about how replays work.

It's certainly not intended that there are no replay control buttons when replaying. The other issue of advancing to the next scenario is already covered in #6180.

Possibly related #4004.

**Comment by Dalas121** [comment]
If I understand correctly, this is about `[change_theme]` overwriting replay controls?

I can verify that this is still an issue on 1.19.11. Some of the new campaigns make heavy use of `[change_theme]` - if this issue is an easy fix it would be great to get it resolved in time for 1.20.

**Comment by soliton-** [comment]
I'd say it's more like the new theme potentially not having any replay controls so none are shown in replay mode.

You could test with a theme that either uses the `REPLAY_THEME` or its contents. `combat-info` theme does for example.

**Comment by Dalas121** [comment]
> I'd say it's more like the new theme potentially not having any replay controls so none are shown in replay mode.

It looks like that's the issue. Would it make sense to ignore `[change_theme]` tags in replays?

**Comment by soliton-** [comment]
Maybe not as a final solution but probably a good idea as a mitigation for the chance of not being able to use replays with `[change_theme]`. Perhaps it should be an option when loading the replay so the default stays but you can fix replays when the issue occurs.

**Comment by Dalas121** [comment]
That sounds good to me. I have no knowledge of C++; do you know a good person I could talk to about making that change?

**Comment by soliton-** [comment]
`[change_theme]` is actually implemented in Lua. With `wesnoth.current.user_is_replaying` (not documented for some reason) you get the replay info.

For adding the option you'd need some C++ knowledge but can probably figure out a lot by checking how "Cancel orders" is implemented for example. Plus a bit more work to store the option value somewhere Lua can access.

Might be an area @CelticMinstrel or @gfgtdf are most familiar with.

**Comment by gfgtdf** [comment]
i'm not sure whether that works, in particular because replays can start at an aribitiary point, if for example the [change_theme] happens in turn 1, and then the user wants to replay from turn 2 the change_theme has already happens it it could break again.


in a similar way, when a user clicks "continue play" during a replay (and the change_theme handled during the replay) the user will probably be stuck in the wrong theme.

**Comment by gfgtdf** [comment]
alternativeley we could add the default replay REPLAY_THEME to all themes by default, by having some 
`[default_replay_theme]{REPLAY_THEME}[/default_replay_theme]` in the game config and then falling back to that if it doesn't fint the replay tag here https://github.com/wesnoth/wesnoth/blob/1.18.0/src/replay_controller.cpp#L94

**Comment by Dalas121** [comment]
Can you elaborate on how the second solution would work, with a default replay theme?

I think we all agree on the desired player experience - replays always use the replay theme, in-game uses some arbitrary set-by-the-campaign-dev theme.

**Comment by Wedge009** [comment]
I'm revisiting this because it's been supposedly marked as a blocker. On the surface, a replay not having controls is obviously not good. But after re-reading the early discussion I see the _specific_ complaint from the original poster was about story scenes that just play automatically, with no normal game-play elements. In that regard, I really do see #7123 as resolving that narrow case because it removes what is essentially a redundant replay of a non-interactive scene.

Then I see the issue being re-opened and renamed on the basis that replays in general (not looking narrowly at story scenes) should have controls available. I concede that is a reasonable issue, but probably should have been separated from the original story scene focus.

That notwithstanding, I'd like to see an example of this occurring in the core campaigns - that would justify the blocker label. If it's only for custom content, then I'm not so sure the blocker label is justified - unless it's intended for near-future campaigns since it was Dalas who asked the question about `[change_theme]`.

**Comment by Dalas121** [comment]
> That notwithstanding, I'd like to see an example of this occurring in the core campaigns - that would justify the blocker label. If it's only for custom content, then I'm not so sure the blocker label is justified - unless it's intended for near-future campaigns since it was Dalas who asked the question about [change_theme].

There's quite a few examples in current mainline, unfortunately. The first scenario of HttT is one such case: the scenario uses `[change_theme]` to hide the UI initially. Once the battle starts, it uses `[change_theme]` to restore the default UI.

**Comment by Wedge009** [comment]
Okay, that's good to know - not that this occurs but that there is justification for its blocker status.

I noticed this older #4004 that also seems to be about `[change_theme]` and missing replay controls.

**Comment by Wedge009** [comment]
Yeah, I see it now - the replay stops when it becomes the player's turn. I reckon what this issue has been turned into and #4004 are the same or at least correlated.

**Comment by gfgtdf** [comment]
ftr i had a pro for this one i [think https://github.com/wesnoth/wesnoth/pull/10416](https://github.com/wesnoth/wesnoth/pull/10416), i currently don't remember what the problem with it was tough

**Comment by gfgtdf** [comment]
would be nice if someone could test 10416

**Comment by Dalas121** [comment]
Unfortunately I don't know how to compile from source, but I agree that it would be great to get this resolved in time for 1.20.
